src/algorithms/nolip.jl

```diff
@@ -0,0 +1,223 @@
+- `gamma=nothing`: stepsize to use, defaults to `1/Lf` if not set (but `Lf` is).
+- `adaptive=false`: forces the method stepsize to be adaptively adjusted.
+- `minimum_gamma=1e-7`: lower bound to `gamma` in case `adaptive == true`.
+- `max_backtracks=20`: maximum number of line-search backtracks.
+- `directions=LBFGS(5)`: strategy to use to compute line-search directions.
+
+# References
+- [1] De Marchi, Themelis, "Proximal gradient algorithms under local Lipschitz
+gradient continuity: a convergence and robustness analysis of PANOC",
+arXiv:2112.13000 (2021).
+"""
+
+Base.@kwdef struct NOLIPIteration{R,Tx,Tf,TA,Tg,TLf,Tgamma,D}
+    f::Tf = Zero()
+    A::TA = I
+    g::Tg = Zero()
+    x0::Tx
+    alpha::R = real(eltype(x0))(0.95)
+    beta::R = real(eltype(x0))(0.5)
+    Lf::TLf = nothing
+    gamma::Tgamma = Lf === nothing ? nothing : (alpha / Lf)
+    adaptive::Bool = gamma === nothing
+    minimum_gamma::R = real(eltype(x0))(1e-7)
+    max_backtracks::Int = 20
+    directions::D = LBFGS(5)
+end
+
+Base.IteratorSize(::Type{<:NOLIPIteration}) = Base.IsInfinite()
+
+Base.@kwdef mutable struct NOLIPState{R,Tx,TAx,TH}
+    x::Tx             # iterate
+    Ax::TAx           # A times x
+    f_Ax::R           # value of smooth term
+    grad_f_Ax::TAx    # gradient of f at Ax
+    At_grad_f_Ax::Tx  # gradient of smooth term
+    gamma::R          # stepsize parameter of forward and backward steps
+    y::Tx             # forward point
+    z::Tx             # forward-backward point
+    g_z::R            # value of nonsmooth term (at z)
+    res::Tx           # fixed-point residual at iterate (= x - z)
+    H::TH             # variable metric
+    tau::R = zero(gamma)
+    x_prev::Tx = similar(x)
+    res_prev::Tx = similar(x)
+    d::Tx = similar(x)
+    Ad::TAx = similar(Ax)
+    x_d::Tx = similar(x)
+    Ax_d::TAx = similar(Ax)
+    f_Ax_d::R = zero(real(eltype(x)))
+    grad_f_Ax_d::TAx = similar(Ax)
+    At_grad_f_Ax_d::Tx = similar(x)
+    z_curr::Tx = similar(x)
+    Az::TAx = similar(Ax)
+    grad_f_Az::TAx = similar(Ax)
+    At_grad_f_Az::Tx = similar(x)
+end
+
+f_model(iter::NOLIPIteration, state::NOLIPState) = f_model(state.f_Ax, state.At_grad_f_Ax, state.res, iter.alpha / state.gamma)
+
+function Base.iterate(iter::NOLIPIteration{R}) where {R}
+    x = copy(iter.x0)
+    Ax = iter.A * x
+    grad_f_Ax, f_Ax = gradient(iter.f, Ax)
+    gamma = iter.gamma === nothing ? iter.alpha / lower_bound_smoothness_constant(iter.f, iter.A, x, grad_f_Ax) : iter.gamma
+    At_grad_f_Ax = iter.A' * grad_f_Ax
+    y = x - gamma .* At_grad_f_Ax
+    z, g_z = prox(iter.g, y, gamma)
+    state = NOLIPState(
+        x=x, Ax=Ax, f_Ax=f_Ax, grad_f_Ax=grad_f_Ax, At_grad_f_Ax=At_grad_f_Ax,
+        gamma=gamma, y=y, z=z, g_z=g_z, res=x-z, H=initialize(iter.directions, x),
+    )
+    if (iter.gamma === nothing || iter.adaptive == true)
+        state.gamma, state.g_z, f_Az, f_Az_upp = backtrack_stepsize!(
```

Comment:
So this initial backtracking is not there in the pseudocode in the paper, do I understand it right?

Reply:
The pseudocode gives special treatment to the case `k=0`. Considering the first iteration, then, the initial backtracking on `gamma` is there indeed.

Reply:
True! Didn’t realize it. Partially related to this: do you think the pseudocode in the manuscript could be rearranged so as to avoid “start with step xyz” and just start with step 1? That just struck me as potentially confusing when I first saw it.